full screen mode implementation for dashboard

stacey-gammon opened the pull request:
Fixes https://github.com/elastic/kibana/issues/9268

Release Note:
You can now enter full screen mode when viewing a dashboard. This hides the Chrome and the top nav bar.  If you have any filters applied, you'll see the filter bar, otherwise that will be hidden as well.  To exit full screen mode, hover over and click the Kibana button on the lower left side of the page, or simple press the `ESC` key.

<img width="1331" alt="screen shot 2017-07-25 at 3 51 15 pm" src="https://user-images.githubusercontent.com/16563603/28590710-2172f6ec-7151-11e7-9eb8-c6b30874930a.png">

alexfrancoeur commented:
@Stacey-Gammon woo!! this PR is fantastic. I saw some comments in slack around showing / hiding the query bar and filter bar on hover and agree that should be in addressed in another PR. I also think that for the first iteration, we should address the most prominent use case of showcasing this dashboard. I can't imagine there will be _much_ need for interaction and querying in full screen mode for this scenario. This is looking great! I have some usability comments below. 

* Is it possible to trigger the browser to go in full screen mode as well when clicking the full screen button? I expected my browser to expand to full screen as well when clicking on this button. Not the end of the world if we can't (progress over perfection) but this is common behavior amongst other dashboarding tools that have this feature
* Can we add keyboard controls here for exiting full screen mode? I was aware of the button on the bottom but also attempted to use the `esc` key to leave full screen mode. Just a though. This may conflict with browser full screen modes as well. It'd be nice if hitting `esc` exited both browser and Kibana full screen mode.
* I see full screen mode in edit mode. It seems odd to me to have that there. I assume this is necessary because when resizing Kibana we automatically expand the panels and not shift the around to fit the screen. Is that correct?
* Somewhat related, I noticed that we persist the side bar space until you adjust the browser size
![jun-12-2017 10-41-02](https://user-images.githubusercontent.com/25181823/27039336-b0cf71b6-4f5b-11e7-9aae-30376b0b70c9.gif)

* It's a bit hard to read "Exit full screen mode" in dark mode. There isn't enough contrast. It'd be good to get @snide's feedback here
<img width="283" alt="screen shot 2017-06-12 at 10 26 46 am" src="https://user-images.githubusercontent.com/25181823/27038683-b2bf8bac-4f59-11e7-8964-b496457283da.png">

Hope this feedback helps! Let me know if you need any other clarification

cjcenizal commented:
This is REALLY COOL. This features makes dashboards feel much more polished, to an extent that is disproportionate to the size of the feature.

I noticed a UX bug. When the dashboard has no panels, clicking "Add" should probably pop you out of full-screen mode so you can see the "Add panel" dropdown:

![image](https://user-images.githubusercontent.com/1238659/27059943-7b7ac56e-4f8e-11e7-8a20-1686d1481683.png)

I also couldn't get the filter and query bar to show up in full screen mode. Am I supposed to be able to do that in this version of the PR? If so, what do I need to do to get them to show?

That said, I like all of @alexfrancoeur's suggestions above.

My only suggestion would be about our approach. I suggest we refine the UX first and then tackle the appearance and behavior of the "Exit full screen mode" button as the very last step. I think we can make some adjustments to make it more consistent with the look-and-feel and I'm sure @snide has thoughts on this too.

Stacey-Gammon commented:
@alexfrancoeur -

> Is it possible to trigger the browser to go in full screen mode as well when clicking the full screen button? I expected my browser to expand to full screen as well when clicking on this button. Not the end of the world if we can't (progress over perfection) but this is common behavior amongst other dashboarding tools that have this feature

It's possible (though sounds like it might be browser specific), but I'm not sure I agree with this one.  I tried in Google Docs, and full screen doesn't automatically pop the browser into full screen mode.  I can see a use case of someone wanting to maximize the browser space, but still be able to quickly tab to other browser windows/desktop icons/etc.

> Can we add keyboard controls here for exiting full screen mode? I was aware of the button on the bottom but also attempted to use the esc key to leave full screen mode. Just a though. This may conflict with browser full screen modes as well. It'd be nice if hitting esc exited both browser and Kibana full screen mode.

Great idea, and actually checking out Google Docs full screen mode for the above, looks like they *only* offer a keyboard way to exit Full screen mode.  An interesting idea if we want to get rid of the button on the bottom.
<img width="781" alt="screen shot 2017-06-13 at 9 56 43 am" src="https://user-images.githubusercontent.com/16563603/27091788-31578f48-502f-11e7-82af-feee70295fe0.png">

> I see full screen mode in edit mode. It seems odd to me to have that there. I assume this is necessary because when resizing Kibana we automatically expand the panels and not shift the around to fit the screen. Is that correct?

I can get rid of it from full screen mode, but I could potentially see the use case, albeit much less common.  Maybe just wanting to move around, resize, or see how it will look in Full Screen mode.

> Somewhat related, I noticed that we persist the side bar space until you adjust the browser size

Will investigate

> It's a bit hard to read "Exit full screen mode" in dark mode. There isn't enough contrast. It'd be good to get @snide's feedback here

Good catch, will look into.

@cjcenizal 

> I noticed a UX bug. When the dashboard has no panels, clicking "Add" should probably pop you out of full-screen mode so you can see the "Add panel" dropdown:

Ah good point, will fix.

> I also couldn't get the filter and query bar to show up in full screen mode. Am I supposed to be able to do that in this version of the PR? If so, what do I need to do to get them to show?

Query bar doesn't show up, filter bar should show up *if* you have a filter applied.
<img width="1775" alt="screen shot 2017-06-13 at 11 58 15 am" src="https://user-images.githubusercontent.com/16563603/27091913-9db4fd1a-502f-11e7-9de4-8322e8c0bb48.png">

We can investigate further options for when to show the query bar (perhaps offer some configurations - would be nice to do the same for embedded mode).  But as a first pass, I think we should punt on it.  The kbn top nav is tightly coupled to the chrome being visible, so the logic would have to be refactored out first.  We'd also have to figure out the UI for the configuration.  So I think those are great ideas, but I don't think it should hold up this more simplified version.

Stacey-Gammon commented:
Fixed the bugs mentioned, will wait to hear about overall styling on the button before fixing the dark theme version.

Stacey-Gammon commented:
jenkins, test this

snide commented on src/core_plugins/kibana/public/dashboard/styles/index.less:
```diff
@@ -2,6 +2,44 @@
+
+.exitFullScreenMode:hover {
+  bottom: 0px;
+  transition: all .5s ease;
```
Here's a cleaner animation for you...


![](https://d3vv6lp55qjaqc.cloudfront.net/items/3V1h382D1q181q2A3B01/Screen%20Recording%202017-06-20%20at%2002.54%20PM.gif?X-CloudApp-Visitor-Id=59773&v=cbfb8c24)

Does the following

* Moves the bug to the top, which is where the controls originally start.
* Uses more reliable vars.
* Improves the transition by
    * Using transform instead of position (much smoother, gpu now takes over).
    * Puts transition on the base selector, now works on on and off hover, instead of just on.
    * Speeds it up. In general, no hover animation should take longer than 250ms.
    * When in doubt use ease-in-out, rather than just ease.

*Note that you should also up the padding to `.dashboard-grid` to 20px if you move it up. In general I've been adding similar increased padding across Kibana, so it's not just for this PR that we need it.

```
.exitFullScreenMode {
  background-color: @globalColorLightestGray;
  border-bottom-left-radius: 4px;
  border-bottom-right-radius: 4px;
  border-top-left-radius: 0;
  border-top-right-radius: 0;
  width: 200px;
  height: 40px;
  left: 0;
  right: 0;
  top: 0;
  margin-left: auto;
  margin-right: auto;
  padding-bottom: 0;
  position: fixed;
  transform: translateY(-26px);
  z-index: 50;
  display: block;
  transition: transform .2s ease-in-out;
  border: solid 1px @globalColorLightGray;

  .kuiIcon {
    display: block;
  }

  &:hover {
    transform: translateY(-1px);
  }
}
```

Stacey-Gammon commented:
Added the new styles (with some slight modifications, including leaving the button on the button after chatting with @snide about the way it looks on the top with filters applied).
<img width="296" alt="screen shot 2017-06-20 at 4 26 18 pm" src="https://user-images.githubusercontent.com/16563603/27354331-3763572c-55d5-11e7-8d36-33a7661328fa.png">
<img width="276" alt="screen shot 2017-06-20 at 4 26 22 pm" src="https://user-images.githubusercontent.com/16563603/27354332-376c8086-55d5-11e7-8657-d5ce0fb1f948.png">

Also addressed dark theme variation:
<img width="1569" alt="screen shot 2017-06-20 at 4 23 34 pm" src="https://user-images.githubusercontent.com/16563603/27354314-254abf62-55d5-11e7-81ad-9865baf2ea6d.png">

And remove Full Screen option from edit mode because it's awkward with the lack of the obvious `Save` option.  Can easily add it in later if we deem it appropriate.

kobelb commented on src/core_plugins/kibana/public/dashboard/dashboard.js:
```diff
@@ -286,7 +287,43 @@ app.directive('dashboardApp', function ($injector) {
+        chrome.setVisible(!fullScreenMode);
+        $scope.$root.$broadcast('globalNav:update');
+        if (fullScreenMode) {
+          onRouteChange = $scope.$on('$routeChangeStart', () => chrome.setVisible(true));
```
I'm not following what we're doing here, we're calling `chrome.setVisible` above.

Stacey-Gammon replied:
I'll add a comment to make it clearer.  I need to add this so that if the user navigates to a different app, or clicks the back button, the user is bounced out of full screen mode.  Otherwise, you can end up with something like this:
<img width="1769" alt="screen shot 2017-06-21 at 2 33 57 pm" src="https://user-images.githubusercontent.com/16563603/27400616-aff0d2dc-568e-11e7-941a-c696dd081c8b.png">

kobelb replied:
Gotcha. When the user navigates away, how is the `onRouteChange`/deregistration called?

Stacey-Gammon replied:
Good call, I think I just need to add a call to `onRouteChange` inside the `$routeChangeStart` handler.  Seems to do the trick.

kobelb commented on src/core_plugins/kibana/public/dashboard/styles/index.less:
```diff
@@ -13,13 +58,12 @@
   display: block;
   margin: 0;
-  padding: 5px;
+  padding: 20px;
```
Did we mean to change this here? It seems to add quite a bit of padding even when not in full-screen mode.

snide replied:
In general I've increased the padding on a bunch of our inner panels (see monitoring...etc) to give the controls some space. This was actually one of the few places I missed in my big K6 PR.

kobelb replied:
It's more accessible for those of us with fat fingers! 😉

Stacey-Gammon commented on src/core_plugins/kibana/public/dashboard/dashboard.js:
```diff
@@ -286,7 +287,43 @@ app.directive('dashboardApp', function ($injector) {
+        $scope.fullScreenMode = fullScreenMode;
+        dashboardState.setFullScreenMode(fullScreenMode);
+        chrome.setVisible(!fullScreenMode);
+        $scope.$root.$broadcast('globalNav:update');
```
yea, I figured the global navigation did update, so it wasn't that bad, though I hate those broadcasts.  I can do `$window.dispatchEvent(new Event('resize'));` to trigger the re-render of the grid, which should also fixe the bug @nreese  discovered.  How do you feel about that?  I don't think there is really a good way to tell gridster to resize directly.  Or maybe I'm missing a more obvious solution...

kobelb replied:
Using the native `resize` event feels equally gross because it's normally fired by the browser itself, and it feels like we'd be hijacking this as well.

Since the `dashboard-grid` is a directive within the `dashboard` we shouldn't have to resort to using a `$scope.$root.$broadcast` and should be able to just call a function that we pass to it if we need to trigger the resize.

Stacey-Gammon commented:
Great call on the back button @kobelb, and nice find @nreese.  Both issues should be fixed now!

Stacey-Gammon commented:
jenkins, test this

alexfrancoeur commented:
This looks great @Stacey-Gammon!

Stacey-Gammon commented:
jenkins, test this

jimgoodwin commented:
@Stacey-Gammon can we get a Release Note: paragraph at the end of the description describing the functionality please...

Stacey-Gammon commented:
done @jimgoodwin!
